Select pre-release collections that are inevitable

webknjaz opened the pull request:
Previously, the resolvelib dependency resolver that we use for managing the collection installations, was set up to disregard pre-releases met in the dependnecy tree unless `--pre` was passed on the CLI by the end-user or the collection didn't have any stable releases published to Galaxy or an Authomation Hub instance.

With this patch, our dependency resolver will start considering pre-releases that are the only way to satisfy requirements met in the dependnecy tree. Such as, when a dependency is requested restricted by a version range that doesn't match any stable releases but can be satisfied by pre-releases in that range. Though, such pre-releases would be disregarded if there are pre-installed collections that already satisfy the requested requirements.

As with many of our previous design considerations, this one is inspired by the Python packaging ecosystem, pip and PEP 440. It is intended to align better with the natural end-user expectations, from the UX perspective.

##### SUMMARY

<!--- Describe the change below, including rationale and design decisions -->
$sbj.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE

<!--- Pick one below and delete the rest -->

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

N/A

webknjaz committed:
Select pre-release collections that are inevitable

Previously, the resolvelib dependency resolver that we use for
managing the collection installations, was set up to disregard
pre-releases met in the dependnecy tree unless `--pre` was passed
on the CLI by the end-user or the collection didn't have any stable
releases published to Galaxy or an Authomation Hub instance.

With this patch, our dependency resolver will start considering
pre-releases that are the only way to satisfy requirements met in
the dependnecy tree. Such as, when a dependency is requested
restricted by a version range that doesn't match any stable releases
but can be satisfied by pre-releases in that range. Though, such
pre-releases would be disregarded if there are pre-installed
collections that already satisfy the requested requirements.

As with many of our previous design considerations, this one is
inspired by the Python packaging ecosystem, pip and PEP 440. It is
intended to align better with the natural end-user expectations, from
the UX perspective.

jborean93 reviewed (commented):
Do you have an example of pip working like this? I tried with a package I knew had a pre-release and wasn't able to see it use the pre-release. For example `pip install 'pypsrp>=0.9.0'` will not install the `1.0.0b1` release. Maybe it's because it's a beta of a major version bump but it would be good to see evidence that this is how Pip works.

To me it doesn't make too much sense to silently bring in a pre-release, it should be explicitly requested either by a pre-release requirement somewhere or with `--pre` in the cli args.

webknjaz commented:
@jborean93

> Do you have an example of pip working like this?

It doesn't and is declared a bug worth fixing: https://github.com/pypa/pip/issues/11888#issuecomment-1479549142. I'm going to make sure it's fixed in pip.

> To me it doesn't make too much sense to silently bring in a pre-release, it should be explicitly requested either by a pre-release requirement somewhere or with `--pre` in the cli args.

https://peps.python.org/pep-0440/#handling-of-pre-releases says it should work like this. And I agree — offering an obvious solution is better then erroring out. Also, `--pre` is for globally allowing pre-releases anywhere in the tree but the behavior this PR introduces allows just enough of pre-releases to fill in the gaps in the tree, preferring stable versions everywhere else.

It'd be weird if a transitive dependency can only be satisfied with pre-releases so it'd require `--pre` effectively forcing the end-user to allow unstable versions for all of their other unrelated dependencies.

But I agree that maybe it's worth printing out a warning when there's no `--pre` and the resolver selects a pre-release.

I talked to @s-hertel on Slack and we agreed that this is how the resolver should behave. And not just because we've drawn a lot of inspiration from Pip and PEP 440 — it's just more intuitive. And I think this is the right thing to do.

jborean93 commented:
I don't wish to stand in the way of this, I just think that using a pre-release without an explicit opt in acknowledgement is not ideal but that's my personal opinion.

> https://peps.python.org/pep-0440/#handling-of-pre-releases says it should work like this.

While it does, it places it under the section

> Dependency resolution tools SHOULD also allow users to request the following alternative behaviours:

This to me says that it should allow this behaviour but not necessarily as the default, just a way to opt into it. The first example here seems to be covered by `--pre` while this proposed PR is trying to solve the second scenario. But like `--pre` maybe this should be covered under a different argument, or `--pre` is expanded to take in a value to define the `--pre` behaviour chosen. It also says it could be an error rather than something that will be resolved normally.

Personally I would wait to merge this behaviour until it is official that `pip` will include it. The PEP is worded in a way that this seems to be opt-in behaviour and even then the ultimately action is still up to the implementation.

s-hertel commented:
> I talked to @s-hertel on Slack and we agreed that this is how the resolver should behave. And not just because we've drawn a lot of inspiration from Pip and PEP 440 — it's just more intuitive. And I think this is the right thing to do.

To be clear, I haven't looked at this in any depth. But we are in agreement that the current behavior does not match the PEP 440.

webknjaz commented:
@jborean93 I'd rather not wait for pip since there's no timeline yet. I think I'd add a warning/error that would depend on a CLI argument but I'm convinced that letting the resolve return something over nothing is a more pleasant UX. Ultimately, the end-users are still responsible for getting their deps in order but in case of the transitive deps, they're dependent on the packagers' decisions and can't do anything about that. And if those packagers depend on pre-release ranges, they probably know what they're doing so I'd argue it's an explicit pre-release request. We still should make it obvious to the end-users, of course.

webknjaz committed:
Emit a warning if depresolver selects pre-releases

This warning is only shown when the end-user doesn't explicitly
request pre-release versions to be considered globally but the
dependency resolver returns some despite that.

webknjaz commented:
> But I agree that maybe it's worth printing out a warning when there's no `--pre` and the resolver selects a pre-release.

@jborean93 I implemented this in https://github.com/ansible/ansible/commit/aaa04eb75e962ca9a6e2c94df12d4b3ad2156662.

Erroring out pending.

webknjaz committed:
Let users forbid collection pre-releases via CLI

Prior to this patch, whenever the dependency resolver resorted to
falling back to selecting pre-release collection candidates to satisfy
the dependency tree requirements, the CLI program would print out a
warning, listing such pre-release collections that ended up being a
part of the dependency tree.

This change allows the end-users to specify
`--forbid-inevitable-pre-releases` command-line option that will turn
the warning into an error. The pre-release candidates are still not
being rejected early during the resolution but are returned in the
tree, making it possible to display the list of pre-releases that are
the only way to satisfy the dependency requirements.

webknjaz commented:
> Erroring out pending.

@jborean93 this is implemented in https://github.com/ansible/ansible/commit/38f1ccc4a89c5f0daa2df1dc075d03d197d7723a, making the PR ready for review.

jborean93 commented:
Ultimately I'm a +0 on this. I can see that this being compliant with PEP 440 but I don't think that people should be creating collections that depend on a pre-release version without them explicitly putting in the pre-release requirement. Having a version flag without a non-prerelease available seems like a package smell to me.

webknjaz commented:
@sivel I'm mostly curious if this is worth including in 2.16 FTR
